nomad/node_endpoint.go

```diff
@@ -213,6 +213,20 @@ func (n *Node) Deregister(args *structs.NodeDeregisterRequest, reply *structs.No
 	if args.NodeID == "" {
 		return fmt.Errorf("missing node ID for client deregistration")
 	}
+	// Look for the node
```

Comment:
Before the `// Verify the arguments` section can you add:

```
	// Check node write permissions
	if aclObj, err := n.srv.ResolveToken(args.AuthToken); err != nil {
		return err
	} else if aclObj != nil && !aclObj.AllowNodeWrite() {
		return structs.ErrPermissionDenied
	}
```

Reply:
We will need to add a test similar to this: https://github.com/hashicorp/nomad/blob/master/nomad/node_endpoint_test.go#L613

What you need to test is that an ACL with node write/management token work and that another token does not.